depth_image_proc/src/nodelets/point_cloud_xyz_radial.cpp

```diff
@@ -0,0 +1,235 @@
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ * 
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the Willow Garage nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ * 
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+#include <ros/ros.h>
+#include <nodelet/nodelet.h>
+#include <image_transport/image_transport.h>
+#include <sensor_msgs/image_encodings.h>
+#include <image_geometry/pinhole_camera_model.h>
+#include <boost/thread.hpp>
+#include <depth_image_proc/depth_traits.h>
+
+#include <sensor_msgs/point_cloud2_iterator.h>
+
+namespace depth_image_proc {
+
+    namespace enc = sensor_msgs::image_encodings;
+
+    class PointCloudXyzRadialNodelet : public nodelet::Nodelet
+    {
+	// Subscriptions
+	boost::shared_ptr<image_transport::ImageTransport> it_;
+	image_transport::CameraSubscriber sub_depth_;
+	int queue_size_;
+
+	// Publications
+	boost::mutex connect_mutex_;
+	typedef sensor_msgs::PointCloud2 PointCloud;
+	ros::Publisher pub_point_cloud_;
+
+	
+	std::vector<double> D_;
+	boost::array<double, 9> K_;
+  
+	int width_;
+	int height_;
+
+	cv::Mat binned;
+  
+	virtual void onInit();
+
+	void connectCb();
+
+	void depthCb(const sensor_msgs::ImageConstPtr& depth_msg,
+		     const sensor_msgs::CameraInfoConstPtr& info_msg);
+
+	// Handles float or uint16 depths
+	template<typename T>
+	void convert(const sensor_msgs::ImageConstPtr& depth_msg, PointCloud::Ptr& cloud_msg);
+    };
+
+    cv::Mat initMatrix(cv::Mat cameraMatrix, cv::Mat distCoeffs, int width, int height, bool radial)
+    {
+	int i,j;
+	int totalsize = width*height;
+	cv::Mat pixelVectors(1,totalsize,CV_32FC3);
```

Comment:
please use templated matrices all over for maintainability: cv::Mat_<cv::Vec3f> pixelVectors(1, totalsize);